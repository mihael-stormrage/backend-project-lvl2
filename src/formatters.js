import _ from 'lodash';

const stylish = (ast, space = 1) => {
  const indent = '  ';
  const lines = _.sortBy(ast, ['name'])
    .reduce((acc, {
      type, name, children, value, nested,
    }) => {
      const getValue = () => (nested ? stylish(children, space + 1) : value);

      if (type === 'unchanged') acc.push(`${indent.repeat(space * 2)}${name}: ${getValue()}`);
      if (type === 'old') acc.push(`${indent.repeat(space * 2 - 1)}- ${name}: ${getValue()}`);
      if (type === 'new') acc.push(`${indent.repeat(space * 2 - 1)}+ ${name}: ${getValue()}`);

      return acc;
    }, []);
  return `{\n${lines.join('\n')}\n${indent.repeat(space * 2 - 2)}}`;
};

export default stylish;
