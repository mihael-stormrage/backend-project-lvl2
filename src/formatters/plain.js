import _ from 'lodash';

const typeStrings = {
  unchanged: () => '',
  removed: (path) => `Property '${path}' was removed\n`,
  added: (path, value) => `Property '${path}' was added with value: ${value}\n`,
  old: (path, value) => `Property '${path}' was updated. From ${value}`,
  new: (_path, value) => ` to ${value}\n`,
};

const flattenByPath = (e) => {
  if (!e.nested) return e;
  const { children, ...item } = e;
  return [item, _.flatMapDeep(children, flattenByPath)];
};

const plain = (ast) => _.sortBy(_.flatMapDeep(ast, flattenByPath), ['path'])
  .map(({
    type, path, value, nested,
  }) => {
    const getValue = () => (nested ? '[complex value]' : (typeof value === 'string' && `'${value}'`) || value);
    const pathString = path?.join('.');

    return typeStrings[type](pathString, getValue());
  }, []).join('');

export default plain;
