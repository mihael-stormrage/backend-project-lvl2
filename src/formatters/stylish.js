import _ from 'lodash';

const stylish = (ast, space = 1) => {
  const indent = '  ';
  const lines = _.sortBy(ast, ['name'])
    .reduce((acc, {
      type, name, children, value, nested,
    }) => {
      const getValue = () => (nested ? stylish(children, space + 1) : value);
      const getLine = () => {
        if (type === 'unchanged') return `${indent.repeat(space * 2)}${name}: ${getValue()}`;
        if (type === 'old' || type === 'removed') return `${indent.repeat(space * 2 - 1)}- ${name}: ${getValue()}`;
        if (type === 'new' || type === 'added') return `${indent.repeat(space * 2 - 1)}+ ${name}: ${getValue()}`;
        throw new Error('Unexpected node type!');
      };
      return [...acc, getLine()];
    }, []);
  return `{\n${lines.join('\n')}\n${indent.repeat(space * 2 - 2)}}`;
};

export default stylish;
