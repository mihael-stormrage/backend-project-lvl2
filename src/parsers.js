import yaml from 'js-yaml';
import path from 'path';
import fs from 'fs';

const getParser = (extension) => {
  switch (extension) {
    case '.json':
      return JSON.parse;
    case '.yml':
    case '.yaml':
      return yaml.load;
    default:
      throw new Error('Unknown file extension');
  }
};

const getData = (file) => {
  const parse = getParser(path.extname(file));
  return parse(fs.readFileSync(file, 'utf8'));
};

export default getData;
