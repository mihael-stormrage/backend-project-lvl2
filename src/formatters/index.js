import stylish from './stylish.js';

const getFormatter = (name) => {
  switch (name) {
    case 'stylish':
      return stylish;
    default:
      throw new Error('Unknown output format');
  }
};

export default getFormatter;
