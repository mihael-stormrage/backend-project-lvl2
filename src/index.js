import { Command } from 'commander/esm.mjs';
import _ from 'lodash';
import getData from './parsers.js';
import getFormatter from './formatters/index.js';
import stylish from './formatters/stylish.js';

const isNested = (v) => typeof v === 'object' && v !== null;

const processChanged = (val, options) => {
  if (val === undefined) return [];
  const result = [];
  const node = {
    ...options, nested: isNested(val),
  };
  if (node.nested) node.children = _.flatMap(val, (v, k) => processChanged(v, { name: k, type: 'unchanged' }));
  else node.value = val;
  result.push(node);
  return result;
};

const makeDiffAst = (obj1, obj2, path = []) => {
  const keys = [...new Set([...Object.keys(obj1), ...Object.keys(obj2)])];
  return keys.flatMap((key) => {
    const val1 = obj1[key];
    const val2 = obj2[key];

    switch (true) {
      case isNested(val1) && isNested(val2):
        return {
          type: 'unchanged', name: key, children: makeDiffAst(val1, val2, [...path, key]), nested: true,
        };
      case val1 === val2:
        return {
          type: 'unchanged', name: key, value: val1, nested: false,
        };
      default: {
        const items = [
          processChanged(val1, { name: key, path: [...path, key], type: 'old' }),
          processChanged(val2, { name: key, path: [...path, key], type: 'new' }),
        ].flat();
        if (items.length === 2) return items;
        const typeCast = { old: 'removed', new: 'added' };
        const [item] = items;
        item.type = typeCast[item.type];
        return items;
      }
    }
  });
};

const genDiff = (data1, data2, formatter = stylish) => {
  if (_.isEqual(data1, data2)) return 'No difference';
  const diff = makeDiffAst(data1, data2);
  return formatter(diff);
};

const getDiff = (file1, file2, formatName) => genDiff(
  getData(file1),
  getData(file2),
  getFormatter(formatName),
);

const program = new Command();

program.name('gendiff').version('1.0.0')
  .description('Compares two configuration files and shows a difference.')
  .arguments('<filepath1> <filepath2>')
  .helpOption('', 'output usage information')
  .option('-f, --format [type]', 'output format', 'stylish')
  .action((file1, file2) => console.log(getDiff(file1, file2, program.opts().format)));

export { getDiff as default, genDiff, program as gendiff };
