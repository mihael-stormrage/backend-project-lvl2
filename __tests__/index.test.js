import fs from 'fs';
import path from 'path';
import genDiff, { gendiff } from '../src';
import getData from '../src/parsers';
import { plain, json } from '../src/formatters';

const getFixturePath = (file) => path.join('__fixtures__', file);
const getFixture = (file) => fs.readFileSync(getFixturePath(file), 'utf-8');
const getFixtureData = (file) => getData(path.join('__fixtures__', file));

let data1;
let data2;

beforeEach(() => {
  data1 = getFixtureData('file1.json');
  data2 = getFixtureData('file2.yml');
});

test('gendiff --help', () => {
  expect(gendiff.helpInformation()).toBe(getFixture('help'));
});

test('genDiff(data1, data2) fn', () => {
  expect(genDiff(data1, data1)).toBe('No difference');
  expect(genDiff(data1, data2)).toBe(getFixture('stylish').trim());
});

test('diff plain format', () => {
  expect(genDiff(data1, data2, plain)).toBe(getFixture('plain'));
});

test('diff in json format', () => {
  expect(genDiff(data1, data2, json)).toMatchSnapshot();
});
