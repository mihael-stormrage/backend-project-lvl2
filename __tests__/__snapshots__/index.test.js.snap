// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff in json format 1`] = `
"[
  {
    \\"type\\": \\"unchanged\\",
    \\"name\\": \\"common\\",
    \\"children\\": [
      {
        \\"type\\": \\"unchanged\\",
        \\"name\\": \\"setting1\\",
        \\"value\\": \\"Value 1\\",
        \\"nested\\": false
      },
      {
        \\"name\\": \\"setting2\\",
        \\"path\\": [
          \\"common\\",
          \\"setting2\\"
        ],
        \\"type\\": \\"removed\\",
        \\"nested\\": false,
        \\"value\\": 200
      },
      {
        \\"name\\": \\"setting3\\",
        \\"path\\": [
          \\"common\\",
          \\"setting3\\"
        ],
        \\"type\\": \\"old\\",
        \\"nested\\": false,
        \\"value\\": true
      },
      {
        \\"name\\": \\"setting3\\",
        \\"path\\": [
          \\"common\\",
          \\"setting3\\"
        ],
        \\"type\\": \\"new\\",
        \\"nested\\": false,
        \\"value\\": null
      },
      {
        \\"type\\": \\"unchanged\\",
        \\"name\\": \\"setting6\\",
        \\"children\\": [
          {
            \\"type\\": \\"unchanged\\",
            \\"name\\": \\"key\\",
            \\"value\\": \\"value\\",
            \\"nested\\": false
          },
          {
            \\"type\\": \\"unchanged\\",
            \\"name\\": \\"doge\\",
            \\"children\\": [
              {
                \\"name\\": \\"wow\\",
                \\"path\\": [
                  \\"common\\",
                  \\"setting6\\",
                  \\"doge\\",
                  \\"wow\\"
                ],
                \\"type\\": \\"old\\",
                \\"nested\\": false,
                \\"value\\": \\"\\"
              },
              {
                \\"name\\": \\"wow\\",
                \\"path\\": [
                  \\"common\\",
                  \\"setting6\\",
                  \\"doge\\",
                  \\"wow\\"
                ],
                \\"type\\": \\"new\\",
                \\"nested\\": false,
                \\"value\\": \\"so much\\"
              }
            ],
            \\"nested\\": true
          },
          {
            \\"name\\": \\"ops\\",
            \\"path\\": [
              \\"common\\",
              \\"setting6\\",
              \\"ops\\"
            ],
            \\"type\\": \\"added\\",
            \\"nested\\": false,
            \\"value\\": \\"vops\\"
          }
        ],
        \\"nested\\": true
      },
      {
        \\"name\\": \\"follow\\",
        \\"path\\": [
          \\"common\\",
          \\"follow\\"
        ],
        \\"type\\": \\"added\\",
        \\"nested\\": false,
        \\"value\\": false
      },
      {
        \\"name\\": \\"setting4\\",
        \\"path\\": [
          \\"common\\",
          \\"setting4\\"
        ],
        \\"type\\": \\"added\\",
        \\"nested\\": false,
        \\"value\\": \\"blah blah\\"
      },
      {
        \\"name\\": \\"setting5\\",
        \\"path\\": [
          \\"common\\",
          \\"setting5\\"
        ],
        \\"type\\": \\"added\\",
        \\"nested\\": true,
        \\"children\\": [
          {
            \\"name\\": \\"key5\\",
            \\"type\\": \\"unchanged\\",
            \\"nested\\": false,
            \\"value\\": \\"value5\\"
          }
        ]
      }
    ],
    \\"nested\\": true
  },
  {
    \\"type\\": \\"unchanged\\",
    \\"name\\": \\"group1\\",
    \\"children\\": [
      {
        \\"name\\": \\"baz\\",
        \\"path\\": [
          \\"group1\\",
          \\"baz\\"
        ],
        \\"type\\": \\"old\\",
        \\"nested\\": false,
        \\"value\\": \\"bas\\"
      },
      {
        \\"name\\": \\"baz\\",
        \\"path\\": [
          \\"group1\\",
          \\"baz\\"
        ],
        \\"type\\": \\"new\\",
        \\"nested\\": false,
        \\"value\\": \\"bars\\"
      },
      {
        \\"type\\": \\"unchanged\\",
        \\"name\\": \\"foo\\",
        \\"value\\": \\"bar\\",
        \\"nested\\": false
      },
      {
        \\"name\\": \\"nest\\",
        \\"path\\": [
          \\"group1\\",
          \\"nest\\"
        ],
        \\"type\\": \\"old\\",
        \\"nested\\": true,
        \\"children\\": [
          {
            \\"name\\": \\"key\\",
            \\"type\\": \\"unchanged\\",
            \\"nested\\": false,
            \\"value\\": \\"value\\"
          }
        ]
      },
      {
        \\"name\\": \\"nest\\",
        \\"path\\": [
          \\"group1\\",
          \\"nest\\"
        ],
        \\"type\\": \\"new\\",
        \\"nested\\": false,
        \\"value\\": \\"str\\"
      }
    ],
    \\"nested\\": true
  },
  {
    \\"name\\": \\"group2\\",
    \\"path\\": [
      \\"group2\\"
    ],
    \\"type\\": \\"removed\\",
    \\"nested\\": true,
    \\"children\\": [
      {
        \\"name\\": \\"abc\\",
        \\"type\\": \\"unchanged\\",
        \\"nested\\": false,
        \\"value\\": 12345
      },
      {
        \\"name\\": \\"deep\\",
        \\"type\\": \\"unchanged\\",
        \\"nested\\": true,
        \\"children\\": [
          {
            \\"name\\": \\"id\\",
            \\"type\\": \\"unchanged\\",
            \\"nested\\": false,
            \\"value\\": 45
          }
        ]
      }
    ]
  },
  {
    \\"name\\": \\"group3\\",
    \\"path\\": [
      \\"group3\\"
    ],
    \\"type\\": \\"added\\",
    \\"nested\\": true,
    \\"children\\": [
      {
        \\"name\\": \\"deep\\",
        \\"type\\": \\"unchanged\\",
        \\"nested\\": true,
        \\"children\\": [
          {
            \\"name\\": \\"id\\",
            \\"type\\": \\"unchanged\\",
            \\"nested\\": true,
            \\"children\\": [
              {
                \\"name\\": \\"number\\",
                \\"type\\": \\"unchanged\\",
                \\"nested\\": false,
                \\"value\\": 45
              }
            ]
          }
        ]
      },
      {
        \\"name\\": \\"fee\\",
        \\"type\\": \\"unchanged\\",
        \\"nested\\": false,
        \\"value\\": 100500
      }
    ]
  }
]"
`;
